docs/flip/1009-optimizer-api.md

```diff
@@ -0,0 +1,541 @@
+
+
+# Motivation
+[motivation]: #motivation
+
+Flax contains some optimizers in the `flax.optim` package, but that list is far
+from exhaustive and ideally we would instead use JAX optimizers from a dedicated
+PyPi package.
+
+DeepMind already has a dedicated library — [Optax] — that implements a wide
+range of interesting optimizers and provides a framework to compose new
+optimizers from reusable gradient transformations.
+
+[Optax]: https://github.com/deepmind/optax
+
+
+# Functional API
+[Functional API]: #functional-api
+
+## Functional Implementation
+[Functional Implementation]: #functional-implementation
+
+An optimizer is defined by two pure functions. Since we will be [using Optax]
+we define the `Optimizer` as a named tuple (instead of a
+`flax.struct.dataclass`):
+
+```python
+# flax/opt/base.py
+
+Params = Any
+Gradients = Any
+OptimizerState = NamedTuple
+
+
+class Optimizer(NamedTuple):
+  init: Callable[[Params], OptimizerState]
+  update: Callable[
+      [Gradients, Params, OptimizerState], Tuple[Params, OptimizerState]]
+```
+
+For example, to implement `flax.optim.Momentum` in this new interface:
+
+```python
+# flax/opt/momentum.py
+
+class MomentumState(OptimizerState):
+  momentum: PyTree
+
+
+def momentum(learning_rate: float, beta: float = 0.9) -> Optimizer:
+
+  def init_fn(params: Params) -> OptimizerState:
+    return MomentumState(momentum=jax.tree_map(jnp.zeros_like, params))
+
+  def update_fn(
+    grads: Gradients, params: Params, state: OptimizerState
+  ) -> Tuple[Params, OptimizerState]:
+    new_momentum = jax.tree_multimap(
+        lambda momentum, grad: momentum * beta + grad,
+        opt_state.momentum, grads)
+    new_params = jax.tree_multimap(
+        lambda param, grad: param - learning_rate * grad,
+        params, grads)
+    return new_params, MomentumState(new_momentum)
+
+  return Optimizer(init_fn, update_fn)
+```
+
+Note that unlike in the [previous API], the function `update_fn()` will be
+called directly in the train loop and must take care of processing the pytrees
+of parameters and optimizer state with a `jax.tree_multimap()`. For more
+complex optimizers this can become a hassle but we can provide a small
+functional helper `@multimap` that takes care of this:
+
+```python
+  def update_fn(
+    grads: Gradients, params: Params, state: OptimizerState
+  ) -> Tuple[Params, OptimizerState]:
+
+    @multimap
+    def inner(param, grad, momentum):
+      new_momentum = momentum * beta + grad
+      new_param = param - learning_rate * grad
+      return new_param, new_momentum
+
+    new_params, new_momentum = inner(params, grads, state.momentum)
+    return new_params, MomentumState(new_momentum)
+```
+
+Remarks:
+
+- We'll call `Optimizer` instances from the functional API `optim`, and refer
+  to `flax.optim.Optimizer` instances from the [previous API] as `optimizer`
+  for the remainder of this document.
+- Compared with the [previous API], this new API does not have the option to
```

Comment:
Just for my understanding: I guess this means that with the new functional API, we are using optax exclusively? I was somehow thinking we want to support both optimizers initially, but I think I misunderstood that. Could you please clarify?

Reply:
We switch to Optax exclusively - reformulated `Summary` to make this clearer

 The `Update Plan` is discussed in the end of the FLIP and ends with `Mark optimizers in flax.optim as deprecated` as a last step.

If you have specific ideas how to make the plan clearer feel free to put some comments in that section.